File: api/routes/mcp.py
import uuid
import datetime
import json
import requests
from flask import Blueprint, request, jsonify, g
from ..db.database import get_db

mcp_bp = Blueprint("mcp", __name__, url_prefix="/api/mcp")

MCP_JSONRPC_VERSION = "2.0"
MCP_PROTOCOL_VERSION = "2024-11-05"


def row_to_dict(row):
    return dict(row) if row else None


def rows_to_list(rows):
    return [dict(row) for row in rows]


def parse_headers(headers_value):
    if not headers_value:
        return {}
    if isinstance(headers_value, dict):
        return headers_value
    try:
        return json.loads(headers_value)
    except Exception:
        return {}


def _mcp_post(url: str, method: str, params: dict, headers: dict, req_id: int = 1):
    """streamable-http transport: POST /mcp with JSON-RPC."""
    resp = requests.post(
        f"{url.rstrip('/')}/mcp",
        json={
            "jsonrpc": MCP_JSONRPC_VERSION,
            "id": req_id,
            "method": method,
            "params": params,
        },
        headers={
            "Content-Type": "application/json",
            "Accept": "application/json, text/event-stream",
            **headers,
        },
        timeout=15,
    )
    resp.raise_for_status()
    # streamable-http can return SSE or plain JSON
    ct = resp.headers.get("Content-Type", "")
    if "text/event-stream" in ct:
        # Parse first data: line from SSE stream
        for line in resp.text.splitlines():
            if line.startswith("data:"):
                return json.loads(line[5:].strip())
        return {}
    return resp.json()


def _test_streamable_http(server_url: str, headers: dict):
    """streamable-http 서버 연결 테스트: initialize → tools/list."""
    init_resp = _mcp_post(server_url, "initialize", {
        "protocolVersion": MCP_PROTOCOL_VERSION,
        "capabilities": {},
        "clientInfo": {"name": "connpass-test", "version": "1.0.0"},
    }, headers, req_id=1)

    if init_resp.get("error"):
        return None, f"initialize 실패: {init_resp['error']}"

    tools_resp = _mcp_post(server_url, "tools/list", {}, headers, req_id=2)
    if tools_resp.get("error"):
        return [], None

    tools = (tools_resp.get("result") or {}).get("tools", [])
    return tools, None


def _test_sse(server_url: str, headers: dict):
    """SSE 서버 연결 테스트: GET /sse 도달 여부만 확인."""
    resp = requests.get(
        f"{server_url.rstrip('/')}/sse",
        headers={"Accept": "text/event-stream", **headers},
        stream=True,
        timeout=10,
    )
    resp.raise_for_status()
    return [], None


@mcp_bp.get("/servers")
def list_servers():
    """MCP 서버 목록 조회."""
    try:
        db = get_db()
        rows = db.execute(
            """
            SELECT id, name, url, transport, enabled, created_at
            FROM mcp_servers
            WHERE user_id = ?
            ORDER BY created_at DESC
            """,
            (g.user_id,),
        ).fetchall()
        return jsonify(rows_to_list(rows))
    except Exception as e:
        return jsonify({"error": str(e)}), 500


@mcp_bp.get("/servers/<server_id>")
def get_server(server_id):
    """MCP 서버 단일 조회."""
    try:
        db = get_db()
        row = db.execute(
            "SELECT id, name, url, transport, headers, enabled, created_at FROM mcp_servers WHERE id = ?",
            (server_id,),
        ).fetchone()
        if row is None:
            return jsonify({"error": "Server not found"}), 404
        result = row_to_dict(row)
        if result.get("headers"):
            try:
                result["headers"] = json.loads(result["headers"])
            except Exception:
                result["headers"] = {}
        return jsonify(result)
    except Exception as e:
        return jsonify({"error": str(e)}), 500


@mcp_bp.post("/servers")
def register_server():
    """MCP 서버 등록."""
    try:
        body = request.get_json(silent=True) or {}
        name = body.get("name")
        url = body.get("url")
        transport = body.get("transport", "streamable-http")
        headers = body.get("headers", {})

        if not name or not url:
            return jsonify({"error": "name and url are required"}), 400
        if transport not in ("streamable-http", "sse"):
            return jsonify({"error": "transport must be 'streamable-http' or 'sse'"}), 400

        server_id = str(uuid.uuid4())
        now = datetime.datetime.utcnow().isoformat()
        headers_json = json.dumps(headers) if headers else None

        db = get_db()
        db.execute(
            """
            INSERT INTO mcp_servers (id, user_id, name, url, transport, headers, enabled, created_at)
            VALUES (?, ?, ?, ?, ?, ?, 1, ?)
            """,
            (server_id, g.user_id, name, url, transport, headers_json, now),
        )
        db.commit()

        row = db.execute(
            "SELECT id, name, url, transport, enabled, created_at FROM mcp_servers WHERE id = ?",
            (server_id,),
        ).fetchone()
        return jsonify(row_to_dict(row)), 201
    except Exception as e:
        return jsonify({"error": str(e)}), 500


@mcp_bp.delete("/servers/<server_id>")
def delete_server(server_id):
    """MCP 서버 삭제."""
    try:
        db = get_db()
        row = db.execute(
            "SELECT id FROM mcp_servers WHERE id = ?", (server_id,)
        ).fetchone()
        if row is None:
            return jsonify({"error": "Server not found"}), 404

        db.execute("DELETE FROM mcp_servers WHERE id = ?", (server_id,))
        db.commit()
        return jsonify({"deleted": server_id})
    except Exception as e:
        return jsonify({"error": str(e)}), 500


@mcp_bp.post("/servers/<server_id>/test")
def test_server(server_id):
    """MCP 서버 연결 테스트 (MCP JSON-RPC 프로토콜 사용)."""
    try:
        db = get_db()
        row = db.execute(
            "SELECT id, name, url, transport, headers FROM mcp_servers WHERE id = ?",
            (server_id,),
        ).fetchone()
        if row is None:
            return jsonify({"error": "Server not found"}), 404

        server = row_to_dict(row)
        headers = parse_headers(server.get("headers"))
        transport = server.get("transport", "streamable-http")

        try:
            if transport == "streamable-http":
                tools, err = _test_streamable_http(server["url"], headers)
            else:
                tools, err = _test_sse(server["url"], headers)

            if err:
                return jsonify({"status": "error", "error": err}), 200

            return jsonify({
                "status": "ok",
                "server_id": server_id,
                "transport": transport,
                "tools": tools,
                "tool_count": len(tools),
            })
        except requests.exceptions.ConnectionError as e:
            return jsonify({"status": "error", "error": f"Connection error: {str(e)}"}), 200
        except requests.exceptions.Timeout:
            return jsonify({"status": "error", "error": "Connection timed out"}), 200
        except requests.exceptions.HTTPError as e:
            sc = e.response.status_code if e.response is not None else 500
            return jsonify({"status": "error", "error": f"HTTP {sc}: {str(e)}"}), 200
        except Exception as e:
            return jsonify({"status": "error", "error": str(e)}), 200

    except Exception as e:
        return jsonify({"error": str(e)}), 500

File: server/routes/mcp.ts
import { Router } from "express";
import db from "../db.js";

const router = Router();
const MCP_PROTOCOL_VERSION = "2024-11-05";

function uid(req: import("express").Request): string {
  return (req.headers["x-user-id"] as string) ?? "default";
}

function parseHeaders(value: string | null | undefined): Record<string, string> {
  if (!value) return {};
  try { return JSON.parse(value); } catch { return {}; }
}

async function mcpPost(url: string, method: string, params: object, headers: Record<string, string>, reqId = 1) {
  const resp = await fetch(`${url.replace(/\/$/, "")}/mcp`, {
    method: "POST",
    headers: { "Content-Type": "application/json", "Accept": "application/json, text/event-stream", ...headers },
    body: JSON.stringify({ jsonrpc: "2.0", id: reqId, method, params }),
    signal: AbortSignal.timeout(15000),
  });
  if (!resp.ok) throw new Error(`HTTP ${resp.status}`);
  const ct = resp.headers.get("content-type") ?? "";
  if (ct.includes("text/event-stream")) {
    const text = await resp.text();
    for (const line of text.split("\n")) {
      if (line.startsWith("data:")) return JSON.parse(line.slice(5).trim());
    }
    return {};
  }
  return resp.json();
}

// GET /api/mcp/servers
router.get("/servers", (req, res) => {
  try {
    const rows = db.prepare(
      "SELECT id, name, url, transport, enabled, created_at FROM mcp_servers WHERE user_id = ? ORDER BY created_at DESC"
    ).all(uid(req));
    res.json(rows);
  } catch (err) {
    res.status(500).json({ error: String(err) });
  }
});

// GET /api/mcp/servers/:id
router.get("/servers/:id", (req, res) => {
  try {
    const row = db.prepare(
      "SELECT id, name, url, transport, headers, enabled, created_at FROM mcp_servers WHERE id = ?"
    ).get(req.params.id) as (Record<string, unknown> & { headers?: string }) | undefined;
    if (!row) return res.status(404).json({ error: "Server not found" });
    if (row.headers) {
      try { row.headers = JSON.parse(row.headers as string) as unknown as string; } catch { row.headers = undefined; }
    }
    res.json(row);
  } catch (err) {
    res.status(500).json({ error: String(err) });
  }
});

// POST /api/mcp/servers
router.post("/servers", (req, res) => {
  try {
    const body = req.body ?? {};
    const { name, url, transport = "streamable-http", headers = {} } = body;
    if (!name || !url) return res.status(400).json({ error: "name and url are required" });
    if (!["streamable-http", "sse"].includes(transport))
      return res.status(400).json({ error: "transport must be 'streamable-http' or 'sse'" });
    const id = crypto.randomUUID();
    const now = new Date().toISOString();
    db.prepare(
      "INSERT INTO mcp_servers (id, user_id, name, url, transport, headers, enabled, created_at) VALUES (?, ?, ?, ?, ?, ?, 1, ?)"
    ).run(id, uid(req), name, url, transport, Object.keys(headers).length ? JSON.stringify(headers) : null, now);
    res.status(201).json(
      db.prepare("SELECT id, name, url, transport, enabled, created_at FROM mcp_servers WHERE id = ?").get(id)
    );
  } catch (err) {
    res.status(500).json({ error: String(err) });
  }
});

// DELETE /api/mcp/servers/:id
router.delete("/servers/:id", (req, res) => {
  try {
    if (!db.prepare("SELECT id FROM mcp_servers WHERE id = ?").get(req.params.id))
      return res.status(404).json({ error: "Server not found" });
    db.prepare("DELETE FROM mcp_servers WHERE id = ?").run(req.params.id);
    res.json({ deleted: req.params.id });
  } catch (err) {
    res.status(500).json({ error: String(err) });
  }
});

// POST /api/mcp/servers/:id/test
router.post("/servers/:id/test", async (req, res) => {
  try {
    const row = db.prepare(
      "SELECT id, name, url, transport, headers FROM mcp_servers WHERE id = ?"
    ).get(req.params.id) as { id: string; name: string; url: string; transport: string; headers?: string } | undefined;
    if (!row) return res.status(404).json({ error: "Server not found" });

    const headers = parseHeaders(row.headers);
    const transport = row.transport ?? "streamable-http";

    try {
      if (transport === "streamable-http") {
        const initResp = await mcpPost(row.url, "initialize", {
          protocolVersion: MCP_PROTOCOL_VERSION,
          capabilities: {},
          clientInfo: { name: "connpass-test", version: "1.0.0" },
        }, headers, 1) as { error?: unknown; result?: { tools?: unknown[] } };

        if (initResp.error) {
          return res.json({ status: "error", error: `initialize 실패: ${JSON.stringify(initResp.error)}` });
        }

        const toolsResp = await mcpPost(row.url, "tools/list", {}, headers, 2) as { error?: unknown; result?: { tools?: unknown[] } };
        const tools = toolsResp.error ? [] : (toolsResp.result?.tools ?? []);
        return res.json({ status: "ok", server_id: req.params.id, transport, tools, tool_count: (tools as unknown[]).length });
      } else {
        // SSE: 연결 가능 여부만 확인
        const resp = await fetch(`${row.url.replace(/\/$/, "")}/sse`, {
          headers: { "Accept": "text/event-stream", ...headers },
          signal: AbortSignal.timeout(10000),
        });
        if (!resp.ok) throw new Error(`HTTP ${resp.status}`);
        return res.json({ status: "ok", server_id: req.params.id, transport, tools: [], tool_count: 0 });
      }
    } catch (err) {
      return res.json({ status: "error", error: String(err) });
    }
  } catch (err) {
    res.status(500).json({ error: String(err) });
  }
});

export default router;
